File: 0x00基本算法/AcWing112.py
from math import inf, sqrt
N = 1006
n, d, p = 0, 0, [[0]*4 for _ in range(N)] # p[i]=[x,y,l,r]
def Radar_Installation():
    global n, d, p
    for i in range(1, n+1): p[i][:2] = list(map(int, input().split()))
    b = True
    for i in range(1, n+1):
        if p[i][1] > d:
            b = False
            break
    if not b:
        print("-1")
        return
    for i in range(1, n+1):
        k = sqrt(d**2 - p[i][1]**2)
        p[i][2:] = p[i][0]-k, p[i][0]+k
    p[1:n+1] = sorted(p[1:n+1], key=lambda x: x[2])
    ans = 0
    pos = -inf
    for i in range(1, n+1):
        if pos < p[i][2]:
            ans += 1
            pos = p[i][3]
        else:
            pos = min(p[i][3], pos)
    print(ans)

def main():
    global n, d
    n, d = map(int, input().split())
    Radar_Installation()

main()

File: 0x00基本算法/AcWing112.cpp
#include <iostream>
#include <cmath>
#include <algorithm>
using namespace std;
const int N = 1006;
const double INF = 0x3f3f3f3f;
int n, d;
struct P {
  int x, y;
  double l, r;
  bool operator < (const P x) const {
    return l < x.l;
  }
} p[N];

void Radar_Installation() {
  for (int i=1; i<=n; i++) cin >> p[i].x >> p[i].y;
  bool b = true;
  for (int i=1; i<=n; i++)
    if (p[i].y > d) {
      b = 0;
      break;
    }
  if (!b) {
    cout << "-1";
    return;
  }
  for (int i=1; i<=n; i++) {
    double k = sqrt((double)d*d - (double)p[i].y*p[i].y);
    p[i].l = p[i].x - k, p[i].r = p[i].x + k;
  }
  sort(p+1, p+n+1);
  int ans = 0;
  double pos = -INF;
  for (int i=1; i<=n; i++)
    if (pos < p[i].l) {
      ans++;
      pos = p[i].r;
    } else {
      pos = min(p[i].r, pos);
    }
  cout << ans;
}

int main() {
  cin >> n >> d;
  Radar_Installation();
  return 0;
}